packages/axelar-wasm-std/src/msg_id/tx_hash_event_index.rs

```diff
@@ -0,0 +1,211 @@
+use core::fmt;
+use std::{fmt::Display, str::FromStr};
+
+use cosmwasm_std::HexBinary;
+use lazy_static::lazy_static;
+use regex::Regex;
+
+use super::Error;
+use crate::{hash::Hash, nonempty};
+
+pub struct HexTxHashAndEventIndex {
+    pub tx_hash: Hash,
+    pub event_index: u32,
+}
+
+impl HexTxHashAndEventIndex {
+    pub fn tx_hash_as_hex(&self) -> nonempty::String {
+        format!("0x{}", HexBinary::from(self.tx_hash).to_hex())
+            .try_into()
+            .expect("failed to convert tx hash to non-empty string")
+    }
+}
+
+const PATTERN: &str = "^(0x[0-9a-f]{64})-(0|[1-9][0-9]*)$";
+lazy_static! {
+    static ref REGEX: Regex = Regex::new(PATTERN).expect("invalid regex");
+}
+
+impl FromStr for HexTxHashAndEventIndex {
+    type Err = Error;
+
+    fn from_str(message_id: &str) -> Result<Self, Self::Err>
+    where
+        Self: Sized,
+    {
+        let caps: Vec<_> = REGEX
+            .captures(message_id)
+            .into_iter()
+            .flat_map(|caps| caps.iter().collect::<Vec<_>>())
+            .collect();
+
+        match caps.as_slice() {
+            [Some(_), Some(tx_id), Some(event_index)] => {
+                let tx_id = tx_id.as_str();
+                let event_index = event_index.as_str();
+
+                Ok(HexTxHashAndEventIndex {
+                    tx_hash: HexBinary::from_hex(&tx_id[2..])
+                        .expect("invalid hex")
+                        .as_slice()
+                        .try_into()
+                        .expect("invalid length tx hash"),
```

Comment:
is it 100% certain that this will never fail?

Reply:
since this function already returns a result, I don't think we gain anything by adding a potential panic instead of just returning an error

Reply:
it should never fail unless there is a bug in our code